lib/tree.rb

```diff
@@ -2,61 +2,210 @@ class TreeNode
+  # Space Complexity: O(m), where m is the height of the tree
   def find(key)
-    raise NotImplementedError
+    if !@root
+      return nil
+    end
+    current = @root
+
+    return move_and_find(key, current)
+  end
+
+  def move_and_find(key, current)
+    if key < current.key
+      if !current.left
+        return nil
+      else
+        move_and_find(key, current.left)
+      end
+    elsif key > current.key
+      if !current.right
+        return nil
+      else
+        move_and_find(key, current.right)
+      end
+    elsif key == current.key
+      return current.value
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n), where n is the number of nodes in the tree
+  # Space Complexity: O(n), where n is the number of nodes in the tree
   def inorder
-    raise NotImplementedError
+    node_array = []
+    if !@root
+      return node_array
+    end
+    current = @root
+
+    node_array = traverse_in_order(node_array, current)
+
+    return node_array
+  end
+
+  def traverse_in_order(array, current_node)
+    if current_node.left
+      traverse_in_order(array, current_node.left)
+    end
+    array.push(pretty_node_object(current_node))
+    if current_node.right
+      traverse_in_order(array, current_node.right)
+    end
+    return array
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n), where n is the number of nodes in the tree
+  # Space Complexity: O(n), where n is the number of nodes in the tree
   def preorder
-    raise NotImplementedError
+    node_array = []
+    if !@root
+      return node_array
+    end
+    current = @root
+
+    node_array = traverse_preorder(node_array, current)
+
+    return node_array
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def traverse_preorder(array, current_node)
+    array.push(pretty_node_object(current_node))
+    if current_node.left
+      traverse_preorder(array, current_node.left)
+    end
+    if current_node.right
+      traverse_preorder(array, current_node.right)
+    end
+    return array
+  end
+
+  # Time Complexity: O(n), where n is the number of nodes in the tree
+  # Space Complexity: O(n), where n is the number of nodes in the tree
   def postorder
-    raise NotImplementedError
+    node_array = []
+    if !@root
+      return node_array
+    end
+    current = @root
+
+    node_array = traverse_postorder(node_array, current)
+
+    return node_array
+  end
+
+  def traverse_postorder(array, current_node)
+    if current_node.left
+      traverse_postorder(array, current_node.left)
+    end
+    if current_node.right
+      traverse_postorder(array, current_node.right)
+    end
+    array.push(pretty_node_object(current_node))
+    return array
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n), where n is the number of nodes in the tree
+  # Space Complexity: O(k + m), where k is the number of leaves in the tree and m is the height of the tree
```

Comment:
Since you encounter each node just once, I would say this is O(n).  The height doesn't really affect it.